dex/pair/src/contexts/add_liquidity.rs

```diff
@@ -0,0 +1,300 @@
+    }
+
+    fn get_second_token_reserve(&self) -> &BigUint<M> {
+        &self.storage_cache.second_token_reserve
+    }
+
+    fn set_lp_token_supply(&mut self, amount: BigUint<M>) {
+        self.storage_cache.lp_token_supply = amount;
+    }
+
+    fn get_lp_token_supply(&self) -> &BigUint<M> {
+        &self.storage_cache.lp_token_supply
+    }
+
+    fn set_initial_k(&mut self, k: BigUint<M>) {
+        self.initial_k = k;
+    }
+
+    fn get_initial_k(&self) -> &BigUint<M> {
+        &self.initial_k
+    }
+
+    fn get_caller(&self) -> &ManagedAddress<M> {
+        &self.caller
+    }
+
+    fn set_output_payments(&mut self, payments: ManagedVec<M, EsdtTokenPayment<M>>) {
+        self.output_payments = payments
+    }
+
+    fn get_output_payments(&self) -> &ManagedVec<M, EsdtTokenPayment<M>> {
+        &self.output_payments
+    }
+
+    fn get_opt_accept_funds_func(&self) -> &OptionalArg<ManagedBuffer<M>> {
+        &self.tx_input.args.opt_accept_funds_func
+    }
+
+    fn get_tx_input(&self) -> &dyn TxInput<M> {
+        &self.tx_input
+    }
+}
+
+impl<M: ManagedTypeApi> TxInputArgs<M> for AddLiquidityArgs<M> {
+    fn are_valid(&self) -> bool {
+        self.first_token_amount_min != 0 && self.second_token_amount_min != 0
+    }
+}
+
+impl<M: ManagedTypeApi> TxInputPayments<M> for AddLiquidityPayments<M> {
+    fn are_valid(&self) -> bool {
+        self.is_valid_payment(&self.first_payment.as_ref())
+            && self.is_valid_payment(&self.second_payment.as_ref())
+    }
+}
+
+impl<M: ManagedTypeApi> AddLiquidityPayments<M> {
+    fn is_valid_payment(&self, payment_opt: &Option<&EsdtTokenPayment<M>>) -> bool {
+        match payment_opt {
+            Some(payment) => {
+                payment.amount != 0
+                    && payment.token_nonce == 0
+                    && payment.token_identifier.is_esdt()
+            }
+            None => false,
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> TxInput<M> for AddLiquidityTxInput<M> {
+    fn get_args(&self) -> &dyn TxInputArgs<M> {
+        &self.args
+    }
+
+    fn get_payments(&self) -> &dyn TxInputPayments<M> {
+        &self.payments
+    }
+
+    fn is_valid(&self) -> bool {
+        self.args_match_payments()
+    }
+}
+
+impl<M: ManagedTypeApi> AddLiquidityTxInput<M> {
+    fn args_match_payments(&self) -> bool {
+        self.min_leq_payment_amount(
+            &self.args.first_token_amount_min,
+            &self.payments.first_payment.as_ref(),
+        ) && self.min_leq_payment_amount(
+            &self.args.second_token_amount_min,
+            &self.payments.second_payment.as_ref(),
+        )
+    }
+
+    fn min_leq_payment_amount(
+        &self,
+        min: &BigUint<M>,
+        payment_opt: &Option<&EsdtTokenPayment<M>>,
+    ) -> bool {
+        match payment_opt {
+            Some(payment) => min <= &payment.amount,
+            None => false,
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> AddLiquidityContext<M> {
```

Comment:
Consider `#[inline]`-ing some of the getter/setter methods. Same for the other contexts.

Reply:
done